src/ripple/app/tx/impl/applySteps.cpp

```diff
@@ -50,371 +50,236 @@
         case ttXCHAIN_COMMIT:
-            return invoke_preflight_helper<XChainCommit>(ctx);
+            return f.template operator()<XChainCommit>();
         case ttXCHAIN_CLAIM:
-            return invoke_preflight_helper<XChainClaim>(ctx);
+            return f.template operator()<XChainClaim>();
         case ttXCHAIN_ADD_CLAIM_ATTESTATION:
-            return invoke_preflight_helper<XChainAddClaimAttestation>(ctx);
+            return f.template operator()<XChainAddClaimAttestation>();
         case ttXCHAIN_ADD_ACCOUNT_CREATE_ATTESTATION:
-            return invoke_preflight_helper<XChainAddAccountCreateAttestation>(
-                ctx);
+            return f.template operator()<XChainAddAccountCreateAttestation>();
         case ttXCHAIN_ACCOUNT_CREATE_COMMIT:
-            return invoke_preflight_helper<XChainCreateAccountCommit>(ctx);
+            return f.template operator()<XChainCreateAccountCommit>();
         default:
-            assert(false);
-            return {temUNKNOWN, TxConsequences{temUNKNOWN}};
+            throw UnknownTxnType(txnType);
+    }
+}
+}  // namespace
+
+// Templates so preflight does the right thing with T::ConsequencesFactory.
+//
+// This could be done more easily using if constexpr, but Visual Studio
+// 2017 doesn't handle if constexpr correctly.  So once we're no longer
+// building with Visual Studio 2017 we can consider replacing the four
+// templates with a single template function that uses if constexpr.
+//
+// For Transactor::Normal
+template <
+    class T,
+    std::enable_if_t<T::ConsequencesFactory == Transactor::Normal, int> = 0>
+TxConsequences
+consequences_helper(PreflightContext const& ctx)
+{
+    return TxConsequences(ctx.tx);
+};
+
+// For Transactor::Blocker
+template <
+    class T,
+    std::enable_if_t<T::ConsequencesFactory == Transactor::Blocker, int> = 0>
+TxConsequences
+consequences_helper(PreflightContext const& ctx)
+{
+    return TxConsequences(ctx.tx, TxConsequences::blocker);
+};
+
+// For Transactor::Custom
+template <
+    class T,
+    std::enable_if_t<T::ConsequencesFactory == Transactor::Custom, int> = 0>
+TxConsequences
+consequences_helper(PreflightContext const& ctx)
+{
+    return T::makeTxConsequences(ctx);
+};
+
+static std::pair<NotTEC, TxConsequences>
+invoke_preflight(PreflightContext const& ctx)
+{
+    try
+    {
+        return with_txn_type(ctx.tx.getTxnType(), [&]<typename T>() {
+            auto const tec = T::preflight(ctx);
+            return std::make_pair(
+                tec,
+                isTesSuccess(tec) ? consequences_helper<T>(ctx)
+                                  : TxConsequences{tec});
+        });
+    }
+    catch (UnknownTxnType const& e)
+    {
+        // Should never happen
+        JLOG(ctx.j.fatal())
+            << "Unknown transaction type in preflight: " << e.txnType;
+        assert(false);
+        return {temUNKNOWN, TxConsequences{temUNKNOWN}};
     }
 }
 
-/* invoke_preclaim<T> uses name hiding to accomplish
```

Comment:
Shouldn't these comments stay?

Reply:
I don't find the comment that helpful - I think the code is clear without it. But I also don't mind leaving it in. Fixed in `eb4547ad04 [fold] Re add comment`